src/EditorFeatures/TestUtilities/StubVsServiceExporter`2.cs

```diff
@@ -0,0 +1,74 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.ComponentModel.Composition;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.VisualStudio;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Threading;
+
+// Import Roslyn.Utilities with an alias to avoid conflicts with AsyncLazy<T>. This implementation relies on
+// AsyncLazy<T> from vs-threading, and not the one from Roslyn.
+using RoslynUtilities = Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Editor.UnitTests;
+
+[Export(typeof(IVsService<,>))]
+[PartCreationPolicy(CreationPolicy.NonShared)]
+[PartNotDiscoverable]
+internal class StubVsServiceExporter<TService, TInterface> : IVsService<TService, TInterface>
+    where TService : class
+    where TInterface : class
+{
+    private readonly AsyncLazy<TInterface> _serviceGetter;
+
+    [ImportingConstructor]
+    [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+    public StubVsServiceExporter(
+        [Import(typeof(SAsyncServiceProvider))] IAsyncServiceProvider2 asyncServiceProvider,
+        JoinableTaskContext joinableTaskContext)
+    {
+        _serviceGetter = new AsyncLazy<TInterface>(() => asyncServiceProvider.GetServiceAsync<TService, TInterface>(joinableTaskContext.Factory, throwOnFailure: true)!, joinableTaskContext.Factory);
+    }
+
+    /// <inheritdoc />
+    public Task<TInterface> GetValueAsync(CancellationToken cancellationToken)
+        => _serviceGetter.GetValueAsync(cancellationToken);
+
+    /// <inheritdoc />
+    public Task<TInterface?> GetValueOrNullAsync(CancellationToken cancellationToken)
+    {
+        var value = GetValueAsync(cancellationToken);
+        if (value.IsCompleted)
+        {
+            return TransformResult(value);
+        }
```

Comment:
I love how clean this is, and really reduces allocations since we already have the nullable versions cached.